File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;

    vector<vector<int>> tmp(rows, vector<int>(columns , 0));
    int cnt = 1;
    for(int i = 0; i<rows;i++){
        for(int j = 0; j<columns; j++){
            tmp[i][j] = cnt;
            cnt++;
        }

    }

    vector<vector<int>> turned = tmp;

    for(auto query : queries){
        int x = query[0] - 1;
        int y = query[1] - 1;
        int nx = query[2] - 1;
        int ny = query[3] - 1;
        int minNum = INT_MAX;
        for(int i = x; i<=nx; i++){
            for(int j = y; j<=ny; j++){
                if((i != x && i != nx) && (j != y && j != ny)) continue;
                if(i == x){
                    if( j != y ) {
                        turned[i][j] = tmp[i][j - 1];
                    }else if (j == y){
                        turned[i][j] = tmp[i+1][j];
                    }
                }else if(i == nx){
                    if(j != ny){
                        turned[i][j] = tmp[i][j+1];
                    }else if (j == ny){
                        turned[i][j] = tmp[i-1][j];
                    }
                }else if(j == y){
                    if(i != nx){
                        turned[i][j] = tmp[i+1][j];
                    }else if( i == nx ){
                        turned[i][j] = tmp[i][j+1];
                    }
                }else if(j == ny){
                    if ( i != x ){
                        turned[i][j] = tmp[i-1][j];
                    }else if(i == x){
                        turned[i][j] = tmp[i][j-1];
                    }
                }
                minNum = min(minNum, turned[i][j]);
            }
        }
        answer.push_back(minNum);
        tmp = turned;
    }
     return answer;
}

File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.java

import java.util.*;

/**
 * row x col
 *  (x1, y1, x2, y2)인 정수 4개로 표현
 *  x1 행 y1 열부터 x2 행 y2 열까지의
 *  영역에 해당하는 직사각형에서 테두리에 있는 숫자들을 한 칸씩 시계방향으로 회전
 */
class Solution {

  int[][] map;
  static List<Integer> ans = new ArrayList<>();
  public int[] solution(int rows, int columns, int[][] queries) {
    int[] answer = {};

    int querySize = queries.length;

    map = new int[rows][columns];
    int num = 1;
    for(int i = 0; i<rows; i++){
      for(int j = 0; j<columns; j++){
        map[i][j] = num;
        num++;
      }
    }

    for(int[] query : queries){
      runQuery(query);
      System.out.println();
    }

    //Collections.sort(ans);

    answer = new int[ans.size()];
    
    for(int i = 0; i< ans.size(); i++){
      answer[i] = ans.get(i);
    }

    return answer;
  }

  void runQuery(int[] query){
    Deque<Integer> q = new ArrayDeque<>();

    int minX = query[0] - 1;
    int minY = query[1] - 1;
    int maxX = query[2] - 1;
    int maxY = query[3] - 1;

    int minimum = Integer.MAX_VALUE;

    for(int i = minY; i<= maxY; i++){
      q.offer(map[minX][i]);
      minimum = Math.min(minimum, map[minX][i]);
    }

    for(int i = minX + 1; i<=maxX; i++){
      q.offer(map[i][maxY]);
      minimum = Math.min(minimum, map[i][maxY]);
    }

    for(int i = maxY - 1; i >= minY; i--){
      q.offer(map[maxX][i]);
      minimum = Math.min(minimum, map[maxX][i]);

    }

    for(int i = maxX - 1; i > minX; i--){
      q.offer(map[i][minY]);
      minimum = Math.min(minimum, map[i][minY]);
    }

    ans.add(minimum);
    q.offerFirst(q.pollLast());

    for(int i = minY; i<= maxY; i++){
      map[minX][i] = q.pollFirst();
    }

    for(int i = minX + 1; i<=maxX; i++){
      map[i][maxY] = q.pollFirst();
    }

    for(int i = maxY - 1; i >= minY; i--){
      map[maxX][i] = q.pollFirst();
    }

    for(int i = maxX - 1; i > minX; i--){
      map[i][minY] = q.pollFirst();
    }
  }

//  public static void main(String[] args) {
//    Solution solution = new Solution();
//    solution.solution(6, 6, new int[][]{{2,2,5,4},{3,3,6,6},{5,1,6,3}});
//  }
}
